### PR TITLE
Writable variable dependency resolution

### LINKED ISSUE #25917: Writable Variable dependency resolution
## Reason
The writable variable interface need better dependency resolution. This is easy for AuxKernels, but may prove more difficult for userobjects...

## Design
We add writable variables to the list returned by `getSuppliedItems` for AuxVariables. For user objects we need to make the ordering of UOs depend not only on UO-UO relationships, but also on coupled and writable coupled variables. This is a bit more iffy.

## Impact
Added robustness. Tag @lindsayad

### PR DESCRIPTION
**Pull request by dschwen**
Closes #25917

### CODE REVIEW

**Comment by dschwen** [comment]
I was most worried about introducing dependencies on variables for userobjects that no userobject supplies (because all variables except for writable coupled variables are not provided by userobjects). That does not seem to be a problem.

File: framework/src/userobjects/UserObject.C [hunk 1]
```diff
@@ -93,7 +93,20 @@ UserObject::UserObject(const InputParameters & parameters)
...
 
+  // insert self
   _supplied_uo.insert(name());
+
+  // add items for dependency resolution of writable coupled variables
+  if (auto coupleable = dynamic_cast<Coupleable *>(this); coupleable)
+  {
+    // add writable coupled variables as supplied
+    for (const auto & var : coupleable->getWritableCoupledVariables())
+      _supplied_uo.insert("Variable/" + var->name());
```
Comment (loganharbour) [inline framework/src/userobjects/UserObject.C:104]: Man, I wish this resolution was done with pointers to MooseObjects instead of strings :(

Comment (dschwen) [reply framework/src/userobjects/UserObject.C:104]: Two issues with that:
1. thread copies of objects. But there might be a way to always get a pointer to the TID=0 object. 
2. things that are not MooseObjects like MaterialProperties

Either way, I completely agree, strings is not the way to go.

**Comment by dschwen** [comment]
Ugh, this breaks `ProjectionAux` because it now looks like a circular dependency loop.

**Comment by dschwen** [comment]
The other problem is that anything that is a coupled variable in the input parameters, is automatically added as a dependency. This is a problem in conjunction with ProjectionAux. Two options

1. Remove the deprecated old interface `writableCoupledValue` (which needs ProjectionAux) completely.
2. Add a separate method in input parameters to register written to variables (and not use addCoupledVar for this purpose anymore)

**Comment by dschwen** [comment]
We still need this

**Pull request closed without merging**